caffe2/mkl/operators/pool_op.cc

```diff
@@ -0,0 +1,188 @@
+    bool RunOnDeviceWithOrderNHWC() override;
+
+    // Input: X
+    // Output: Y
+private:
+    vector<TIndex> cached_maxpool_input_dims_;
+    vector<TIndex> cached_avgpool_input_dims_;   
+    LayoutWrapper<T> workspace_layout_;
+    T *workspace_buffer_ = nullptr;
+    PrimitiveWrapper<T> primitive_;
+    MKLMemory<T> buffer_;
+    void* resources_[dnnResourceNumber] = {0};  
+
+};    
+
+template <>
+bool MKLPoolOp<float, MaxPool>::RunOnDeviceWithOrderNCHW() {
+
+    auto& X = OperatorBase::Input<MKLMemory<float>>(0);
+    MKLMemory<float>* Y = OperatorBase::Output<MKLMemory<float>>(0);
+
+    if (cached_maxpool_input_dims_ != X.dims())  {
+        cached_maxpool_input_dims_ = X.dims();
+        
+        // We will utilize the SetOutputSize() function in the base class
+        // with dummy TensorCPU input and output to calculate the sizes.
+        TensorCPU dummy_input(X.dims());
+        TensorCPU dummy_output;
+
+        ConvPoolOpBase<MKLContext>::SetOutputSize(dummy_input,
+                                                  &dummy_output, 
+                                                  X.dim32(1));               
+        size_t dim = X.ndim();
+        CAFFE_ENFORCE(4 == dim);
+
+        int paddings[4] = {-pad_l_, -pad_t_, -pad_r_, -pad_b_};
+        size_t strides[2] = {stride_w_, stride_h_};
+        size_t kernel_size[2] = {kernel_w_, kernel_h_};
+        
+        // Create main primitive.
+        primitive_.Reset(dnnPoolingCreateForward_F32, 
+                            nullptr, 
+                            dnnAlgorithmPoolingMax, 
+                            X.layout(),
+                            kernel_size,
+                            strides,
+                            paddings,
+                            dnnBorderZerosAsymm);
+
+       Y->Reset(dummy_output.dims(), primitive_, dnnResourceDst);
+       buffer_.Reset(dummy_output.dims(), primitive_, dnnResourceDst, true); 
+
+       workspace_layout_.Reset(primitive_, dnnResourceWorkspace);
+       MKLDNN_SAFE_CALL(mkl::dnnAllocateBuffer<float>((void **)(&workspace_buffer_),
+                                                         workspace_layout_));    
+
+    } 
+
+    // Try to share from the output: this allows us to avoid unnecessary copy
+    // operations, if the output is already allocated and is having the same
+    // layout as the buffer has.
+    buffer_.ShareFrom(*Y);
+    resources_[dnnResourceSrc] = X.buffer();
+    resources_[dnnResourceDst] = buffer_.buffer();
+    resources_[dnnResourceWorkspace] = workspace_buffer_;
+    MKLDNN_SAFE_CALL(mkl::dnnExecute<float>(primitive_, resources_));
+    buffer_.CopyTo(Y, primitive_, dnnResourceDst);
+    return true;
+}
+
+template <>
+bool MKLPoolOp<float, AveragePool>::RunOnDeviceWithOrderNCHW() {
+
+    auto& X = OperatorBase::Input<MKLMemory<float>>(0);
+    MKLMemory<float>* Y = OperatorBase::Output<MKLMemory<float>>(0);
+
+    if (cached_avgpool_input_dims_ != X.dims())  {
```

Comment:
this block seems like a lot of duplication with previous one, just algorithm is different. Did you consider having a helper which takes an algorithm as a parameter?